frame/support/src/migrations.rs

```diff
@@ -209,3 +211,69 @@ impl<P: Get<&'static str>, DbWeight: Get<RuntimeDbWeight>> frame_support::traits
 		Ok(())
 	}
 }
+
+/// A migration that can proceed in multiple steps.
+pub trait SteppedMigration {
+	fn id(&self) -> BoundedVec<u8, ConstU32<32>>;
+
+	/// Try to migrate as much as possible with the given weight.
+	///
+	/// **ANY STORAGE CHANGES MUST BE ROLLED-BACK BY THE CALLER UPON ERROR.** This is necessary
+	/// since the caller cannot return a cursor in the error case. `Self::transactional_step` is
+	/// provided as convenience for a caller. A cursor of `None` implies that the migration is at
+	/// its end. TODO: Think about iterator `fuse` requirement.
+	fn step(
+		&self,
+		cursor: &Option<SteppedMigrationCursor>,
+		meter: &mut WeightMeter,
+	) -> Result<Option<SteppedMigrationCursor>, SteppedMigrationError>;
+
+	/// Same as [`Self::step`], but rolls back pending changes in the error case.
+	fn transactional_step(
+		&self,
+		mut cursor: Option<SteppedMigrationCursor>,
+		meter: &mut WeightMeter,
+	) -> Result<Option<SteppedMigrationCursor>, SteppedMigrationError> {
+		with_transaction_opaque_err(move || match self.step(&cursor, meter) {
+			Ok(new_cursor) => {
+				cursor = new_cursor;
+				sp_api::TransactionOutcome::Commit(Ok(cursor))
+			},
+			Err(err) => sp_api::TransactionOutcome::Rollback(Err(err)),
+		})
+		.map_err(|()| SteppedMigrationError::Failed)?
+	}
+}
+
+/// The maximal length in bytes of a cursor for a stepped migration.
+pub const STEPPED_MIGRATION_CURSOR_LEN: u32 = 1024;
+
+/// An opaque cursor that defines the "position" or a migration.
+pub type SteppedMigrationCursor =
+	crate::BoundedVec<u8, crate::traits::ConstU32<STEPPED_MIGRATION_CURSOR_LEN>>;
```

Comment:
so any stepped migration's cursor must be less than this? not sure why we have to enforce it as one global for everyone. I presume better would be to let the pallet decide on this. 

If used on the relay chain, you reasonable might want to set this to `u32::MAX`. On a parachain, in order to have good pre-dispatch PoV, you probably less.